Copy directory instead of cloning an entire repository

ibarrae opened the pull request:
A proposal for closing #116.

If this makes sense, could you let me know @juanpaucar, so I can carry on with some tests for backward compatibility and updating the documentation :smile:

juanpaucar commented on src/System/Hapistrano/Types.hs:
```diff
@@ -57,15 +60,28 @@ data Config =
 
+-- | The origin of the repository. It can be from a version control provider
+-- like GitHub or a local directory.
+data Repository
```
A better name for this could be `Source`

juanpaucar commented on src/System/Hapistrano/Types.hs:
```diff
@@ -57,15 +60,28 @@ data Config =
+-- | The origin of the repository. It can be from a version control provider
+-- like GitHub or a local directory.
+data Repository
+  = ExternalRepository
```
Since we could add support for other sources, this constructor and its fields could be better named after Git ...

ibarrae replied:
I think we could take this in a separate PR, since that involves other changes.

juanpaucar replied:
Since this PR is introducing these new constructors, then it could be modified now. On second thought, `GitRepository` might be a better name for this constructor and a similar approach could be followed for the fields on the record

juanpaucar commented on app/Config.hs:
```diff
@@ -97,8 +97,11 @@ instance FromJSON Config where
-    configRevision   <- o .: "revision"
+        getRepository m =
+              LocalDirectory <$> m .: "local_directory"
+          <|> (m .: "external-repository" >>= \m' -> ExternalRepository <$> m' .: "url" <*> m' .: "revision")
```
I don't think this should be necessary for now and should be safe to remove it. At least not until adding support for more sources like mercury, etc.

ibarrae replied:
Which one you mean it's not necessary? The `external-repository` option?

juanpaucar replied:
I think we can safely remove this whole line until we add support for more sources

ibarrae commented:
Hey @juanpaucar, I've added some tests and also updated the changelog/readme. I bumped the version to `0.4.0.0` on these files but retained from changing in the `.cabal` in case this is wrong. Let me know your thoughts about this :smile:

juanpaucar commented on README.md:
```diff
@@ -50,10 +53,16 @@ restart_command: systemd restart my-app-service
-  specify it as `origin/branch_name` due to the way that the cache repo is
-  configured.
+* Related to the `source` of the repository, you have the following options:
+  - _Git repository_ — consists of two parameters. When these are set,
```
Please mark this as default

juanpaucar commented on spec/System/HapistranoConfigSpec.hs:
```diff
@@ -0,0 +1,53 @@
+          , targetSshArgs = []
+          }
+      ]
+    , configSource = LocalDirectory { localDirectoryPath = $(mkAbsDir "/") }
```
Why is the default configuration for the test the `local` revision?  I think it should be the other way around. And judging by the parser, that is the case. Am I missing something?

ibarrae replied:
You are totally right, thanks

juanpaucar commented on src/System/Hapistrano/Config.hs:
```diff
@@ -97,8 +97,10 @@ instance FromJSON Config where
           (maybeToList (Target <$> host <*> pure port <*> pure shell <*> pure sshArgs) ++ hs)
-    configRepo       <- o .: "repo"
-    configRevision   <- o .: "revision"
+        getRepository m =
```
this might not just be `getRepository` but source

ibarrae commented:
Thanks for your feedback @juanpaucar, I bumped the version to `0.4.0.0` and also added a changelog entry. I can't think of anything else to add in this PR. Let me know what you think of that.
